virtual_rainforest/models/abiotic_simple/simple_regression.py

```diff
@@ -0,0 +1,588 @@
+    output["atmospheric_pressure"] = data["atmospheric_pressure"].where(
+        data["atmospheric_pressure"].coords["layer_roles"] == "soil",
+        (data["atmospheric_pressure_ref"].isel(time=time_index)),
+    )
+
+    # Mean atmospheric C02 profile, [ppm]
+    output["atmospheric_co2"] = data["atmospheric_co2"].where(
+        data["atmospheric_co2"].coords["layer_roles"] == "soil",
+        (data["atmospheric_co2_ref"].isel(time=time_index)),
+    )
+
+    # Precipitation at the surface is reduced as a function of leaf area index
+    precipitation_surface = data["precipitation"].isel(time=time_index) * (
+        1 - water_interception_factor * data["leaf_area_index"].sum(dim="layers")
+    )
+
+    # Mean soil moisture profile, [] and Runoff, [mm]
+    soil_moisture_only, surface_run_off = calculate_soil_moisture(
+        layer_roles=layer_roles,
+        precipitation_surface=precipitation_surface,
+        current_soil_moisture=data["soil_moisture"],
+        soil_moisture_capacity=soil_moisture_capacity,
+    )
+    output["soil_moisture"] = xr.concat(
+        [
+            data["soil_moisture"].isel(
+                layers=np.arange(0, len(layer_roles) - layer_roles.count("soil"))
+            ),
+            soil_moisture_only,
+        ],
+        dim="layers",
+    )
+    output["surface_run_off"] = surface_run_off
+
+    return output
+
+
+# supporting functions
+def logarithmic(x: DataArray, gradient: float, intercept: float) -> DataArray:
+    """Logarithmic function.
+
+    Args:
+        x: x values
+        gradient: gradient
+        intercept: intercept
+
+    Returns:
+        y values
+    """
+    for number in x:
+        if number < 0:
+            to_raise = ValueError("x values must be positive!")
+            LOGGER.error(to_raise)
+            raise to_raise
+
+    return DataArray(gradient * np.log(x) + intercept)
+
+
+def log_interpolation(
+    data: Data,
+    reference_data: DataArray,
+    leaf_area_index_sum: DataArray,
+    layer_roles: List,
+    layer_heights: DataArray,
+    upper_bound: float,
+    lower_bound: float,
+    gradient: float,
+) -> DataArray:
+    """LAI regression and logarithmic interpolation of variables for vertical profile.
+
+    Args:
+        data: Data object
+        reference_data: input variable at reference height
+        leaf_area_index_sum: leaf area index summed over all layers, [m m-1]
+        layer_roles: list of layer roles (soil, surface, subcanopy, canopy, above)
+        layer_heights: vertical layer heights, [m]
+        lower_bound: minimum allowed value
+        upper_bound: maximum allowed value
+        gradient: gradient of regression from :cite:t:`hardwick_relationship_2015`
+
+    Returns:
+        vertical profile of provided variable
+    """
+
+    # Calculate microclimatic variable at 1.5 m as function of leaf area index
+    lai_regression = DataArray(
+        leaf_area_index_sum * gradient + reference_data, dims="cell_id"
+    )
```

Comment:
Oh right, because otherwise it changes from `numpy` to `DataArray` and the return type is already fixed. It's things like this that I find a little frustrating about `mypy`!

Reply:
oh yes, mypy...